add migratory connectivity code to birdflow

adokter opened the pull request:
MigConnectivity package does not use memory efficiently, bringing MC calculation into birdflow

adokter committed:
bring migratory connectivity code inside birdflow

adokter committed:
fix typo

adokter committed:
avoid matrix kronecker product

adokter committed:
add informative text messages about progress

Commit:
remove test postfix

adokter committed:
remove postfix

Commit:
Merge branch 'fix_mc' of https://github.com/birdflow-science/BirdFlowExtras into fix_mc

adokter committed:
add capitals

Commit:
Merge branch 'fix_mc' of https://github.com/birdflow-science/BirdFlowExtras into fix_mc

adokter committed:
add more vectorization

adokter committed:
add parallel computation

adokter committed:
remove parallelism, seems unnecessary

adokter committed:
remove test postfix in function name

Commit:
Merge branch 'fix_mc' of https://github.com/birdflow-science/BirdFlowExtras into fix_mc

Commit:
Use element-wise multiplication, slightly more efficeint

adokter committed:
minor typo fix

ethanplunkett commented on R/calc_birdflow_mc.R:
```diff
@@ -41,11 +42,28 @@ calc_birdflow_mc <- function(bf, ...) {
+  sd_V <- sqrt(sum((target_dist - mu_V)^2 * target_abun_prod))
+
+  # multiply transition matrix and relative abundance
+  psi_abun <- apply(psi, 2, "*", origin_abun)
```
This is the same as:
```
psi_abun <- psi * origin_abun  
``` 
which will be faster.

adokter replied:
fixed

adokter committed:
base target abundances on origin distribution times transition matrix

adokter committed:
fix target distribution calculation

ethanplunkett committed:
test for consistent output from calc_birdflow_mc()

ethanplunkett committed:
update test

adokter committed:
take distributions from marginals instead of the observed distributions

adokter committed:
reformulation of calc_route_mc

adokter committed:
update documentation and add references

HermioneDeng committed:
Use slice to prevent finding several nearest timestep.
Add nsteps in nearest_timestep function so we can define the minimum timesteps it needs to have away from the tstep.

adokter commented on R/calc_route_mc.R:
```diff
@@ -66,19 +67,21 @@ calc_route_mc <- function(rts, bf, exact=TRUE, ...) {
   else{
-    origin <- nearest_timestep(rts, origin_t)
-    target <- nearest_timestep(rts, target_t)
+    origin <- nearest_timestep(rts, origin_t, nsteps = 10)
```
@HermioneDeng I would expose this `nsteps` argument also in the function itself, so add another argument before the `...` and document the new argument in the header.

HermioneDeng committed:
change to delta_steps and expose it to the function

HermioneDeng committed:
Change delta_steps default value. Add parameter in the document.

HermioneDeng committed:
fix the route_id matching bug

ethanplunkett committed:
lint

ethanplunkett committed:
Bump version. Update tests. Pass Check.
